src/shared/components/modal/createModalLauncher.tsx

```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import { Modal, ModalProps } from '@patternfly/react-core';
+
+export type ModalComponentProps = Omit<ModalProps, 'children' | 'isOpen' | 'appendTo'> & {
+  ref?: React.LegacyRef<Modal>;
+};
+
+export type CreateModalLauncher = <P extends ModalComponentProps>(
+  C: React.ComponentType<P>,
+  args?: {},
+) => (props: P) => (onClose: () => void) => React.ReactElement;
```

Comment:
Add a proper type for `(onClose: () => void) => React.ReactElement;` and use the same in ModalContext